presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -480,13 +508,19 @@ public boolean equals(Object obj)
             MultiJoinNode other = (MultiJoinNode) obj;
             return this.sources.equals(other.sources)
                     && ImmutableSet.copyOf(extractConjuncts(this.filter)).equals(ImmutableSet.copyOf(extractConjuncts(other.filter)))
-                    && this.outputSymbols.equals(other.outputSymbols);
+                    && this.outputSymbols.equals(other.outputSymbols)
+                    && this.pushedProjectionThroughJoin == other.pushedProjectionThroughJoin;
         }
 
-        static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Lookup lookup, int joinLimit)
+        static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Context context, boolean pushProjectionsThroughJoin)
+        {
+            return toMultiJoinNode(joinNode, context.getLookup(), context.getIdAllocator(), getMaxReorderedJoins(context.getSession()), pushProjectionsThroughJoin);
+        }
+
+        static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Lookup lookup, PlanNodeIdAllocator planNodeIdAllocator, int joinLimit, boolean pushProjectionsThroughJoin)
         {
             // the number of sources is the number of joins + 1
-            return new JoinNodeFlattener(joinNode, lookup, joinLimit + 1).toMultiJoinNode();
+            return new JoinNodeFlattener(joinNode, lookup, planNodeIdAllocator, joinLimit + 1, pushProjectionsThroughJoin).toMultiJoinNode();
```

Comment:
I would refactor this code, to:

```
return new JoinNodeFlattener(lookup, planNodeIdAllocator, joinLimit + 1).toMultiJoinNode(joinNode, pushProjectionsThroughJoin);
```

To me it is awkward that all processing is happening in the constructor. I would call `flattenNode` from  `toMultiJoinNode`. Then  `flattenNode` could have some context passed recursively (or return some value) hence then you wouldn't need to have fields `pushProjectionsThroughJoin` and `pushedProjectionThroughJoin` as they would be maintained during the call of `toMultiJoinNode` only.